h2/src/main/org/h2/security/auth/DefaultAuthenticator.java

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright 2004-2018 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: Alessandro Ventura
+ */
+package org.h2.security.auth;
+
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.xml.bind.JAXB;
+
+import org.h2.api.CredentialsValidator;
+import org.h2.api.UserToRolesMapper;
+import org.h2.engine.Database;
+import org.h2.engine.Right;
+import org.h2.engine.Role;
+import org.h2.engine.SysProperties;
+import org.h2.engine.User;
+import org.h2.engine.UserBuilder;
+import org.h2.message.Trace;
+import org.h2.security.auth.impl.AssignRealmNameRole;
+import org.h2.security.auth.impl.JaasCredentialsValidator;
+import org.h2.util.StringUtils;
+
+/**
+ * Default authenticator implementation.
+ * <p>
+ * When client connectionInfo contains property AUTHREALM={realName} credentials
+ *  (typically user id and password) are validated by
+ *  by {@link org.h2.api.CredentialsValidator} configured for that realm.
+ * </p>
+ * <p> 
+ * When client connectionInfo doesn't contains AUTHREALM property credentials
+ *  are validated internally on the database
+ * </p> 
+ * <p>
+ * Rights assignment can be managed through {@link org.h2.api.UserToRolesMapper}
+ * </p>
+ * <p>
+ * Default configuration has a realm H2 that validate credentials through JAAS api (appName=h2).
+ * To customize configuration set h2.authConfigFile system property to refer a valid h2auth.xml config file
+ * </p> 
+ */
+public class DefaultAuthenticator implements Authenticator {
+
+    public static final String DEFAULT_REALMNAME = "H2";
+
+    private Map<String, CredentialsValidator> realms = new HashMap<>();
+
+    private List<UserToRolesMapper> userToRolesMappers = new ArrayList<>();
```

Comment:
the above two fields need to be marked as final to prevent Tomcat's classloader from messing with them

Reply:
That works too